README.md
# 项目配置  
npm install  
在项目根目录下添加  .screeps.json  
```json
{
  "branch":"default",
  "email": "your email in screeps",
  "password": "your password"

}
```
执行grunt推上代码  
# feature 
2019/9/10 自杀系统  
2019/9/27 cacheController 能够自动将多segment合并为一个 来保存大于100kb的对象  
2019/9/27 war module 从控制台进行指挥的战争系统  
# bugfixd  
2019/9/9 修复了power-b在治疗时左右横跳的bug  
2019/9/9 修复了watcher堵门口的bug  
2019/9/10 修改powercreep的移动cost=1
2019/9/13 修复creep对穿roomPosition越界问题
2019/9/13 在有视野的敌对区域执行正常寻路

# 进攻日志  
2019/9/5 10235956 进攻E14N41 玩家guy  
2019/9/6 10241074 进攻E14N41失败 对方填充tower优先级过高  
2019/9/9          nuke E36N49 E33N38  
2019/9/9 10349375 进攻E14N41 玩家guy 
2019/9/10         进攻E14N41失败 敌人开启了safemode   
2019/9/24 nuke E15N41 overmind-1  
2019/9/24 nuke E34N47 玩家Meekohi  
2019/9/24 nuke E27N47 玩家Devnix  
2019/9/27 nuke E25N48 玩家Devnix  
2019/9/27 attack E23N47 Blokkie_NL  
2019/9/27 attack E31N41 barret50cal  
2019/9/27 attack E23N41 dandykong  
2019/9/27 attack E17N39 TwelveBaud  
2019/9/28 nuke E19N39 Morkonena  
2019/9/28 attack E22N42 Ower  
2019/9/28 attack E19N45 Azusa  
2019/9/28 attack E38N41 TKerstiens  
2019/9/28 attack E34N47 Meekohi  
2019/9/28 attack E17N42 woofmao  
2019/9/28 attack E31N32 nullpointerat  
2019/9/28 attack E32N33 fredboy  
2019/9/28 attack E33N37 damfle  
2019/9/28 attack E36N42 GCZ  
2019/9/28 attack E34N42 grgushi  
2019/9/28 attack E33N41 Cardy31  
2019/9/30 nuke E32N32 AndrewTawin  
2019/9/30 nuke E21N43 thdetdestro  
2019/9/30 nuke E19N47 Devnix  
2019/9/30 attack E12N41 ags131  
2019/10/1 nuke E33N38 hackgpp    
2019/10/2 nuke E16N38 TwelveBaud      
2019/10/3 nuke E21N39 Devnix  
2019/10/5 attack E11N39 ags131  
2019/10/5 attack E33N36 hackgpp 
2019/10/6 nuke E39N47 Armaos 
2019/10/6 attack E33N38 hackgpp  
Game.war.init(['E12N41',[39,2,'E12N40'],{SEAL:1},['E14N41'],{SEAL:{smallattack:true}}])  
Game.war.init(['E11N38',[24,47,'E11N39'],{SEAL:1},['E14N41'],{SEAL:{smallattack:true}}])  
2019/10/8 nuke E37N38 IC4U  
Game.war.init(['E12N41',[39,2,'E12N40'],{SEAL:1},['E14N41'],{SEAL:{smallattack:true}}])  
Game.war.init(['E14N39',[24,46,'E14N40'],{SEAL:1},['E14N41'],{SEAL:{smallattack:true}}])  
Game.war.init(['E12N38',[30,46,'E12N39'],{SEAL:1},['E14N41'],{SEAL:{smallattack:true}}])  
Game.war.init(['E11N37',[4,31,'E12N37'],{SEAL:1},['E14N41'],{SEAL:{smallattack:true}}])  
Game.war.init(['E12N36',[4,31,'E12N37'],{SEAL:1},['E14N41'],{SEAL:{smallattack:true}}])  
Game.war.init(['E8N41',[4,6,'E9N41'],{SEAL:1},['E14N41'],{SEAL:{smallattack:true}}])  
Game.war.init(['E7N41',[41,2,'E7N40'],{SEAL:1},['E14N41'],{SEAL:{smallattack:true}}])  
Game.war.init(['E9N45',[4,26,'E10N45'],{SEAL:1},['E14N41'],{SEAL:{smallattack:true}}])  
Game.war.init(['E31N45',[48,25,'E30N45'],{SEAL:1},['E28N46'],{SEAL:{wall:true}}])  
Game.war.init(['E41N39',[43,24,'E40N39'],{boostAttack:1,boostHeal:1},['E27N38'],{},['boost']])  
Game.war.init(['E30N35',[25,25,'E30N35'],{boostAttack:1,boostHeal:1},['E27N38'],{},['boost','keep']])  
Game.war.init(['E43N39',[40,47,'E43N40'],{boostAttack:1,boostHeal:1},['E27N38'],{},['boost']])  
Game.war.init(['E40N38',[25,25,'E40N38'],{boostAttack:1,boostHeal:1},['E27N38'],{},['boost']])  
Game.war.init(['E42N34',[25,25,'E40N39'],{SEAL:1},['E27N38'],{SEAL:{heal:true}}])  
Game.war.init(['E18N44',[11,27,'E19N44'],{SEAL:1},['E19N41'],{SEAL:{smallattack:true}}])
Game.war.init(['E33N38',[29,46,'E33N39'],{SEAL:1},['E29N38'],{SEAL:{smallattack:true}}])
Game.war.init(['E12N38',[30,46,'E12N39'],{SEAL:1},['E14N41'],{SEAL:{smallattack:true}}])  
Game.war.init(['E6N41',[39,44,'E6N42'],{SEAL:1},['E14N41'],{SEAL:{smallattack:true}}])  
Game.war.init(['E16N43',[13,2,'E16N42'],{SEAL:1},['E14N41'],{SEAL:{heal:true}}])  
Game.war.init(['E38N39',[44,27,'E37N39'],{SEAL:1},['E29N41'],{SEAL:{heal:true}}])  
Game.war.init(['E11N38',[24,47,'E11N39'],{SEAL:2},['E14N41'],{SEAL:{heal:true}}])  
Game.war.init(['E13N42',[14,2,'E13N41'],{SEAL:1},['E14N41'],{SEAL:{heal:true}}])  
Game.war.init(['E14N39',[24,46,'E14N40'],{SEAL:1},['E14N41'],{SEAL:{smallattack:true}}])  
Game.war.init(['E22N39',[20,7,'E22N38'],{SEAL:1},['E19N41'],{SEAL:{smallattack:true}}])  
Game.war.init(['E24N38',[42,23,'E23N38'],{SEAL:1},['E19N41'],{SEAL:{smallattack:true}}])  
Game.war.init(['E6N49',[21,24,'E7N50'],{SEAL:2},['E14N41'],{SEAL:{heal:true}}])  
Game.war.init(['E9N36',[15,9,'E9N35'],{SEAL:1},['E14N41'],{SEAL:{smallattack:true}}])  
Game.war.init(['E21N34',[32,2,'E21N33'],{SEAL:2},['E19N41'],{SEAL:{heal:true}}])  
Game.war.init(['E23N34',[3,14,'E24N34'],{SEAL:2},['E27N38'],{SEAL:{heal:true}}])  
Game.war.init(['E25N44',[18,3,'E25N43'],{boostAttack:1,boostHeal:1},['E25N43'],{},['boost','keep']])  
Game.war.init(['E8N48',[4,15,'E9N48'],{SEAL:1},['E14N41'],{SEAL:{smallattack:true}}])  
Game.war.init(['E16N38',[22,31,'E16N39'],{SEAL:1},['E14N41'],{SEAL:{smallattack:true}}])  
Game.war.init(['E16N43',[15,10,'E16N42'],{SEAL:1},['E14N41'],{SEAL:{smallattack:true}}])  
Game.war.init(['E11N39',[25,25,'E11N40'],{SEAL:1},['E14N41'],{SEAL:{smallattack:true}}])  
Game.war.init(['E6N49',[21,24,'E7N50'],{SEAL:1},['E14N41'],{SEAL:{heal:true}}])  
Game.war.init(['E8N35',[16,3,'E8N34'],{SEAL:1},['E14N41'],{SEAL:{heal:true}}])  

Game.war.init(['E32N44',[45,28,'E31N44'],{boostAttack:1,boostHeal:1},['E28N46'],{},['boost','keep']])  
Game.war.init(['E33N48',[24,16,'E33N48'],{boostAttack:1,boostHeal:1},['E28N46'],{},['boost','keep']])  
Game.war.init(['E32N44',[45,28,'E31N44'],{boostHeal:1},['E28N46'],{},['boost','keep']])  
Game.war.init(['E37N49',[13,45,'E37N50'],{boostHeal:1},['E28N46'],{},['boost','keep']])  
Game.war.init(['E18N44',[11,27,'E19N44'],{SEAL:1},['E19N41'],{SEAL:{smallattack:true}}])
Game.war.init(['E29N37',[15,48,'E9N38'],{boostHeal:1},['E14N41'],{boostHeal:{heallv6:true}},['boost','keep']])  
Game.war.init(['E34N42',[29,3,'E34N41'],{boostAttack:1,boostHeal:1},['E29N41'],{boostHeal:{heallv7:true},boostAttack:{attacklv7:true}},['boost','keep']])  
Game.war.init(['E38N41',[32,3,'E38N40'],{boostHeal:1},['E29N41'],{boostHeal:{heallv6:true}},['boost','keep']])  

Game.war.init(['E25N44',[17,2,'E25N43'],{boostAttack:1,boostHeal:1},['E25N43'],{boostHeal:{heallv6:true},boostAttack:{attacklv6:true}},['boost','keep']])  
Game.war.init(['E14N44',[43,2,'E14N42'],{boostAttack:1,boostHeal:1},['E14N41'],{
boostHeal:{'tough': 13,'move': 10,'heal': 27},
boostAttack: {'tough': 15,'move': 10,'ranged_attack': 25}},['boost','keep']])  
Game.war.init(['E24N43',[47,27,'E24N44'],{boostAttack:1,boostHeal:1},['E25N43'],{boostHeal:{heallv7:true},boostAttack:{attacklv7:true}},['boost','keep']])  
Game.war.init(['E22N44',[14,2,'E22N43'],{boostAttack:1,boostHeal:1},['E19N41'],{boostHeal:{heallv7:true},boostAttack:{attacklv7:true}},['boost','keep']])  
Game.war.init(['E27N43',[31,47,'E27N44'],{boostAttack:1,boostHeal:1},['E25N43'],{boostHeal:{heallv7:true},boostAttack:{attacklv7:true}},['boost','keep']])  

CHANGELOG.md
2020-02-12
==========

  * 春节大更新
    builder 只修墙和rampart会出 一个小的
    控制台交易对话系统
    动态每日价格
    自动刷墙 背包刷墙
    自动定时1000tick紫球
    工厂原矿自动压缩
    存入memory的ext顺序表
    添加optower
    大修墙link反向输出
    miner寻路到 range1
    nuke自动顶住 pc逃跑
    观察terminal
    re'action自动反应
    seal看守模式
    t3化合物自动平均
    化合物共享
    tower 5 1分打
    计算effect的tower伤害
    todo:战斗Pc

2020-01-22
==========

  * Update CHANGELOG.md
  * bug fix
    添加性能统计
    搬运工无目标禁止出生
    4级及以下filler完全填充,出生后计算extension填充顺序
    Game.memory.js segment可以直接写入
    pc添加oplab 优化regensource逻辑
    bucket分层工作
    miner修好container之后主动计算任务
    矿工container满了停止挖掘
    opener直到修好4级storage spawn extension才停止出生
    订单hack JSON
    反应自动上限
    修路优先storage中心
    grafana添加creep pc性能统计
    push代码的tick不进行solveextension

2020-01-21
==========

  * prototype.SpeedUp.getAllOrders
    自动规划4向旋转
    carryer近距离不忽略creep
    pc满载bug自动释放
    自动规划矿井
    opener提前miss
    pc使用名字硬编码room
    bug:商店胡乱挂单 未修复

2020-01-20
==========

  * upgrader safemode
    提高linkmanager填充效率
    upgrader空位寻路
    safemode警觉boost目标
  * 更改tower策略 打人后break

2020-01-19
==========

  * add tower selecter
  * fix tower heal solve bug
  * add whiteList
  * add hits grafana
    添加墙厚度的可视化
  * Merge branch 'master' of https://github.com/Yuandiaodiaodiao/Screeps
  * del
  * Update README.md
  * update auto complete

2020-01-16
==========

  * changevisual

2020-01-09
==========

  * add opener fix market
    未完成的自动规划
    grafana支持

2019-11-18
==========

  * bugfix

2019-10-21
==========

  * backup for new memoryController

2019-10-20
==========

  * updataFactory

2019-10-18
==========

  * Merge branch 'master' of https://github.com/Yuan-ZH/Screeps
  * add
  * fix war normal

2019-10-15
==========

  * ?

2019-10-13
==========

  * boost attack

2019-10-08
==========

  * boost

2019-09-30
==========

  * 重构spawn

2019-09-27
==========

  * add more

2019-09-11
==========

  * add

2019-09-09
==========

  * add

2019-09-03
==========

  * terminal auto sell

2019-08-28
==========

  * Merge branch 'master' of https://github.com/Yuan-ZH/Screeps
  * add autocomplere
  * add
  * dig power

2019-08-20
==========

  * add

2019-07-29
==========

  * centerroom

2019-07-25
==========

  * double rooms

2019-07-22
==========

  * add reserver

2019-07-21
==========

  * add default

2019-07-18
==========

  * Initial commit


src/main.js
Game.lodash = require('lodash-my')


require('prototype.SpeedUp.getAllOrders')
load()

require('prototype.Creep.move')
// require('prototype.Find.cache')
require('prototype.Whitelist')
require('prototype.Room.structures')
var tower = require('tower')
var link = require('link')
console.log('reload---------------------')
require('command')

function clearmem() {
    for (let name in Memory.creeps) {
        if (!Game.creeps[name]) {
            delete Memory.creeps[name]
        }
    }
    for (let name in Memory.powerCreeps) {
        if (!Game.powerCreeps[name] || !Game.powerCreeps[name].ticksToLive) {
            delete Memory.powerCreeps[name]
        }
    }
}

module.exports.roomjson = {
    ctm: [],
    subroom: [],
    source: [],
    missions: {},
    tower: [],
    mineral: [],
    link: []
}

var missions_ori = {
    filler: {},
    watcher: {},
    centerminer: {},
    subprotecter: {},
    miner: {//挖矿&修复附近的建筑物
    },
    linkmanager: {},
    carryer: {//from A transfer to B
    },
    rattacker: {},

    coreminer: {},


    upgrader: {},
    reserver: {},


    builder: {},
    flagworker: {},
    collecter: {},
    mineraler: {},
    opener: {},
    terminalmanager: {},
    farcarryer: {},
    destroyer: {},
}


function mission_generator(room) {

// Memory.rooms[room.name].missions=missions
    if (room.spawns.length === 0) return
    room.memory.missions = room.memory.missions || {}
    let thisroom = Memory.rooms[room.name]
    thisroom.missions = require('tools').deepcopy(missions_ori)
    //找到所有资源
    const sources = []
    let targets = Game.tools.findrooms(room, FIND_SOURCES)
    for (let obj of targets) {
        sources.push(obj)
    }
    //修路
    try {
        if (Game.runTime > 0) {
            Game.RoomPlanner.autoRoad(room)
        }

    } catch (e) {
        console.log('buildroad error' + e + 'in' + room.name)
    }
    //找矿
    const minerals = []
    {
        const mineralstemp = room.find(FIND_MINERALS)
        if (mineralstemp.length > 0) {
            for (let obj of mineralstemp) {
                if (obj.pos.findInRange(FIND_STRUCTURES, 1,
                    {filter: obj => obj.structureType === STRUCTURE_EXTRACTOR}).length > 0) {
                    minerals.push(obj)
                } else {
                    //建造矿井
                    room.createConstructionSite(obj.pos, STRUCTURE_EXTRACTOR)
                    room.createConstructionSite(Game.tools.nearavailable(obj.pos), STRUCTURE_CONTAINER)
                }
            }
        }
    }
    {
        //centerlink
        if (room.storage) {
            let centerlink = room.storage.pos.findClosestByRange(FIND_STRUCTURES, {filter: obj => obj.structureType == STRUCTURE_LINK})
            if (centerlink) thisroom.centerlink = centerlink.id
            thisroom.wallLink = []
            for (let link of room.links) {
                if (link.pos.findInRange(FIND_SOURCES, 2).length > 0) {
                    continue
                }
                if (link.pos.x <= 3 || link.pos.y <= 3 || link.pos.x >= 45 || link.pos.y >= 45) {
                    continue
                }
                if (link.id != thisroom.centerlink) {
                    thisroom.wallLink.push(link.id)
                }
            }
            if (thisroom.wallLink.length === 0) {
                thisroom.wallLink = undefined
            }
        }
    }
    //findlab
    if (room.controller.level == 8) {
        thisroom.lab = {}
        try {
            let labs = room.labs
            if (labs.length == 10) {
                let xb = 0
                let yb = 0
                for (let lab of labs) {
                    xb += lab.pos.x
                    yb += lab.pos.y
                }
                xb /= 10
                yb /= 10
                labs.sort((a, b) => Math.abs(a.pos.x - xb) + Math.abs(a.pos.y - yb) - Math.abs(b.pos.x - xb) - Math.abs(b.pos.y - yb))
                thisroom.lab['input'] = [labs[0].id, labs[1].id]
                thisroom.lab['output'] = []
                labs.slice(2).forEach(obj => thisroom.lab['output'].push(obj.id))
                thisroom.lab['ok'] = true
            }
        } catch (e) {
            console.log('findlab' + e)
        }
    }

    require('tools').solveExtension(room)
    let missions = thisroom.missions
    //分配miner
    if (room.controller.level >= 4) {
        for (let source of sources) {
            const container = source.pos.findInRange(FIND_STRUCTURES, 1, {
                filter: structure => structure.structureType == STRUCTURE_CONTAINER
            })[0]
            const link = source.pos.findInRange(FIND_STRUCTURES, 2, {
                filter: structure => structure.structureType == STRUCTURE_LINK && structure.my == true
            })[0]
            if (room.controller.level < 8) {
                if (link && container) {
                    container.destroy()
                }
            }
            missions.miner[source.id] = {
                target: source.id,
                container: container ? container.id : undefined,
                link: link ? link.id : undefined,
            }

        }
    }


    //分配carryer
    for (let sourceid in missions.miner) {
        if (missions.miner[sourceid].container) {
            if (Game.getObjectById(missions.miner[sourceid].link)) continue
            let canfill = room.links
            let containerid = missions.miner[sourceid].container
            let container = Game.getObjectById(containerid)
            let filltarget = null
            let mincost = 999
            if (!room.storage) {
                let spawns = room.spawns
                if (spawns.length == 0) continue
                let spawn = spawns[0]
                filltarget = spawn.pos.findClosestByRange(FIND_STRUCTURES, {
                    filter: obj => obj.structureType === STRUCTURE_CONTAINER
                })
                if (filltarget)
                    mincost = PathFinder.search(container.pos, {pos: filltarget.pos, range: 1}, {
                        plainCost: 2, swampCost: 10, roomCallback: require('tools').roomc_nocreep
                    }).cost
            } else {

                filltarget = room.storage
                mincost = PathFinder.search(container.pos, {pos: filltarget.pos, range: 0}, {
                    plainCost: 2, swampCost: 10, roomCallback: require('tools').roomc_nocreep
                }).cost
                if (container.pos.roomName !== room.name) {
                    for (let obj of canfill) {
                        let nowcost = PathFinder.search(container.pos, {pos: obj.pos, range: 1}, {
                            plainCost: 2, swampCost: 10, roomCallback: require('tools').roomc_nocreep
                        }).cost
                        if (nowcost <= mincost) {
                            filltarget = obj
                            mincost = nowcost
                        }
                    }
                }

            }

            if (filltarget) {
                if (mincost > 1 || room.controller.level === 4) {
                    missions.carryer[containerid] = {
                        gettarget: containerid,
                        fill: filltarget.id,
                        type: RESOURCE_ENERGY,
                        carrycost: mincost,
                    }
                }

            }

        }
    }

    //分配upgrader
    let controller = room.controller
    missions.upgrader[controller.id] = {
        controller: controller.id,
    }

    //分配filler
    {
        let storage = room.storage
        if (storage) {

            missions.filler = {}
            missions.filler[storage.id] = {
                gettarget: storage.id,
            }
        } else {
            let spawns = room.find(FIND_MY_SPAWNS)
            if (spawns.length > 0) {
                let spawn = spawns[0]
                let gettarget = spawn.pos.findClosestByRange(FIND_STRUCTURES, {
                    filter: obj => obj.structureType == STRUCTURE_CONTAINER
                })
                if (gettarget) {
                    missions.filler[gettarget.id] = {
                        gettarget: gettarget.id,
                    }
                }

            }

        }
    }


    //分配watcher
    for (let subroom of thisroom.subroom) {
        missions.watcher[subroom] = {
            roomName: subroom,
        }
    }


    //opener
    //mineraler
    if (room.controller.level >= 6) {
        for (let source of minerals) {

            const container = source.pos.findInRange(FIND_STRUCTURES, 3, {
                filter: structure => structure.structureType == STRUCTURE_CONTAINER
            })[0]
            if (container && !source.ticksToRegeneration) {
                //已经配置了container 允许采矿
                missions.mineraler[source.id] = {
                    target: source.id,
                    container: container.id,
                }
            }
        }
    }


    //mineraler配套carryer
    for (let sourceid in missions.mineraler) {
        if (missions.mineraler[sourceid].container) {
            let containerid = missions.mineraler[sourceid].container
            let filltarget = room.terminal
            let source = Game.getObjectById(sourceid)
            if (filltarget && !source.ticksToRegeneration) {
                missions.carryer[containerid] = {
                    gettarget: containerid,
                    fill: filltarget.id,
                    type: source.mineralType,
                }
            }

        }
    }


    // //rattacker
    // missions.rattacker[room.name] = {
    //     roomName: room.name
    // }
    // farcarryer
    // missions.farcarryer[room.name] = {
    //     roomName: room.name
    // }
    // // controllerattack
    // missions.controllerattack[room.name] = {
    //     roomName: room.name
    // }


    //linkmanager
    {
        const storage = room.storage
        if (storage) {
            const container = room.storage.pos.findInRange(FIND_STRUCTURES, 6, {filter: obj => obj.structureType == STRUCTURE_CONTAINER}).sort(
                (a, b) => {
                    return a.pos.getRangeTo(room.controller.pos) - b.pos.getRangeTo(room.controller.pos)
                }
            )[0]
            const link = room.storage.pos.findInRange(FIND_STRUCTURES, 6, {filter: obj => obj.structureType == STRUCTURE_LINK})[0]
            if (container && link) {
                missions.linkmanager[room.name] = {
                    roomName: room.name,
                    link: link.id,
                    container: container.id
                }
            }
        }
    }


    for (let miss in missions) {
        if (_.isEmpty(missions[miss])) {
            delete missions[miss]
        }
    }
    missions['subprotecter'] = missions['subprotecter'] || {}
}


let lastTime = 0

function timer(strs = null) {
    let timeuse = Game.cpu.getUsed() - lastTime
    if (true && strs) console.log(strs + "  " + timeuse.toFixed(4))
    lastTime = Game.cpu.getUsed()
    return timeuse
}

function shard() {
    if (Game.shard.name != 'shard3') {
        module.exports.loop = function () {
        }
        return true
    }
    return false
}

Memory.cpu.pushTime = Game.time

function load() {
    Game.test = require('test').test
    Game.war = require('war')
    Game.config = require('config')
    Game.missionController = require('missionController')
    Game.tools = require('tools')
    Game.terminal = require('terminal')
    Game.factory = require('factory')
    Game.observer = require('observer')
    Game.RoomPlanner = require('RoomPlanner')
    Game.lodash = require('lodash-my')
    // require('prototype.SpeedUp.getAllOrders').load()
    Game.runTime = Game.time - Memory.cpu.pushTime
    Game.defend = require('defendController')
    Game.reaction = require('reaction')

}

var missionController = require('missionController')
// profiler.enable()
module.exports.loop = function () {
    if (shard()) return

    load()
    try {
        require('Game.memory').work()
    } catch (e) {
        console.log('main.Game.memory error' + e)
    }
    // try {
    //     require('cacheController').Cache()
    // } catch (e) {
    //     console.log('cacheController error' + e)
    // }
    require('prototype.Creep.move').clear()
    if (Game.time % 20 == 0) {
        clearmem()
    }
    if (Game.time % 1000 == 0) {
        require('observer').find()
        // Game.memory.roomCache = {}
        for (let roomName in Memory.rooms) {
            try {

                mission_generator(Game.rooms[roomName])
            } catch (e) {
                console.log(roomName + 'mission_generator error ' + e)
            }
        }

    }

    if (Game.time % 50 === 0) {
        for (let roomName in Memory.rooms) {
            const room = Game.rooms[roomName]
            try {
                require('defendController').miss(room)

            } catch (e) {
                console.log('subprotecter error' + e)
            }
        }
    }
    if (Game.time % 10 == 0) {

        for (let roomName in Memory.rooms) {
            const room = Game.rooms[roomName]
            try {
                require('subprotecter').miss(room)

            } catch (e) {
                console.log('subprotecter error' + e)
            }
        }
    }
    if (Game.time % 10 == 0) {
        try {
            require('nuke').work()
        } catch (e) {
            console.log('nuke.work error' + e)
        }
        try {
            missionController.detector()
        } catch (e) {
            console.log(`missionController.detector ${e}`)
        }
        for (let roomName in Memory.rooms) {
            let room = Game.rooms[roomName]
            try {
                missionController.spawner(room)
            } catch (e) {
                console.log('mission_spawner' + e)
            }
        }
    }
    {
        try {
            require('observer').work()
        } catch (e) {
            console.log('observerwork' + e)
        }

    }
    if ((Game.time - 20) % 50 === 0) {
        for (let roomName in Game.config.obterminal) {
            try {
                require('terminalObserver').work(roomName)
            } catch (e) {
                console.log('terminal ob error' + e)
            }
        }


    }
    for (let roomName in Memory.rooms) {
        let room = Game.rooms[roomName]
        if (!room) continue
        if (Game.time % 2 === 0) {
            Game.factory.doReact(room)
        }
        try {
            missionController.spawn(room)
        } catch (e) {
            console.log('missionController.spawn' + e)
        }
        try {
            require('roomvisual').work(room)
        } catch (e) {
            console.log('roomvisual' + roomName + e)
        }
        try {
            require('spawn').work(room)
        } catch (e) {
            console.log('spawnerror' + e)
        }

        try {
            link.work(room)
        } catch (e) {
            console.log(roomName + 'link error ' + e)
        }

        if (room.towers.length === 0 && room.find(FIND_HOSTILE_CREEPS).length > 0 && (room.controller.level <= 3 || room.controller.level >= 8)) {
            room.controller.activateSafeMode()
        }

        try {
            const powers = room.powerSpawn
            if (powers && powers.power > 0) {
                powers.processPower()
            }
        } catch (e) {
            console.log('pows ' + e)
        }
    }

    if ((Game.time + 20) % 500 == 0) {
        try {
            require('observer').cache()
        } catch (e) {
            console.log('observer' + e)
        }
    }

    timer()
    Object.values(Game.powerCreeps).forEach(obj => {
        try {
            if (Game.time <= Game.rooms[obj.name].memory.runAwayTick && Game.rooms[obj.name].memory.runAwayTick - Game.time <= 100) {
                require('nukeWall').pcRunAway(obj)
            } else {
                require('powerscreep').work(obj)
            }
        } catch (e) {
            console.log('powerscreep ' + obj.name + e)
        }
    })
    Memory.cpu.pcCpu = timer()
    if ((Game.time + 25) % 50 == 0) {
        try {
            require('powerBank').cache()
        } catch (e) {
            console.log('powerBank-cache' + e)
        }
    }
    if (Game.time % 50 == 0) {

        try {
            require('powerBank').miss()
        } catch (e) {
            console.log('powerBank-miss' + e)
        }
        for (let roomn in Memory.powerPlan) {
            try {
                require('powerBank').solveplan(roomn)
            } catch (e) {
                console.log('powerBank-solveplan' + e)
            }
        }
    }

    if (Game.time % 10 === 0) {
        for (let roomName in Memory.rooms) {
            let room = Game.rooms[roomName]
            let ticks = 100
            let rate = room.storage ? room.storage.store[RESOURCE_ENERGY] / room.storage.store.getCapacity() : 0
            if (room.storage && rate >= 0.9) {
                ticks = Math.ceil(Math.round(Math.max(10, 50 - (0.93 - 0.9) * 800) / 10)) * 10
            }
            if (Game.time % ticks === 0) {
                try {
                    require('terminal').work(room)
                } catch (e) {
                    console.log('terminal' + roomName + e)
                }
            }

        }
    }
    if ((Game.time - 25) % 100 === 0) {
        try {
            Game.terminal.avgT3()
        } catch (e) {
            console.log('main avgT3 error' + e)
        }
    }
    if ((Game.time - 24) % 100 === 0) {
        try {
            Game.terminal.sendOverT3()
        } catch (e) {
            console.log('main sendOverT3 error' + e)
        }
    }
    if ((Game.time - 13) % 100 === 0) {
        try {
            Game.terminal.fillOverT3()
        } catch (e) {
            console.log('main sendOverT3 error' + e)
        }
    }
    if (Game.time % 100 === 0) {


        try {
            require('opener').miss()
        } catch (e) {
            console.log('opener.miss ' + e)
        }

        //terminal
        for (let roomName in Memory.rooms) {
            let room = Game.rooms[roomName]
            try {
                require('wallWorker').miss(room)
            } catch (e) {
                console.log('wallWorker.miss error' + e)
            }
            try {
                require('nukeWall').miss(room)
            } catch (e) {
                console.log('nukeWall.miss error' + e)

            }
            try {
                Game.factory.miss(room)
            } catch (e) {
                console.log('Game.factory.miss error' + e)
            }
            try {
                require('reaction').work(room)
            } catch (e) {
                console.log('reaction' + roomName + e)
            }
            try {
                require('upgrader').miss(room)

            } catch (e) {
                console.log('upgrader miss' + roomName + e)
            }
            try {
                require('builder').miss(room)
            } catch (e) {
                console.log('builder miss' + e)
            }
            try {
                require('terminalmanager').miss(room)
            } catch (e) {
                console.log('terminalmanager' + e)
            }
            try {
                require('reserver').miss(room)
            } catch (e) {
                console.log('reserver-miss' + e)
            }
        }
    }

    if (Game.time % 5 == 0) {
        for (let roomName in Memory.rooms) {
            let room = Game.rooms[roomName]
            try {
                require('reaction').doreaction(room)
            } catch (e) {
                console.log('doreaction' + roomName + e)
            }
        }
    }
    timer()
    Object.values(Game.creeps).forEach(obj => {
        try {
            if (!obj.spawning) {
                const type = obj.name.split('_')[1]
                let time = Game.cpu.getUsed()
                if (type === 'wallWorker' || type === 'warCarry' || type === 'warWall') {
                    if (!((Game.cpu.bucket < 9000 && Object.keys(Memory.powerPlan).length > 0) || Game.cpu.bucket < 3000)) {
                        require(type).work(obj)
                    }
                } else if (type === 'upgrader') {
                    if (Game.cpu.bucket > 3000) {
                    }
                } else if (type === 'carryer') {
                    if (Game.cpu.bucket > 2000) {
                        require(type).work(obj)
                    }
                } else if (type === 'boostAttack' || type === 'boostHeal' || type === 'subprotecter' || type === 'destroyer' || type === 'SEAL') {
                    require(type).work(obj)
                } else if (Game.cpu.bucket > 1000) {
                    require(type).work(obj)
                } else if (Game.time % 3 !== 0) {
                    require(type).work(obj)
                }
                let ticks = Game.cpu.getUsed() - time
                if (ticks > Memory.highTicks) {
                    console.log(`${Game.time % 1000} high cpu in ${obj.name}  cost=${ticks.toFixed(3)}`)
                }
            }
        } catch (e) {
            console.log('role=' + obj.name + 'error' + e)
        }
    })
    Memory.cpu.creepCpu = timer()


    if (Game.time % 20 === 0) {
        Game.war.miss()
    }

    if (Game.market.credits > 2e6) {
        let tokens = Game.market.getAllOrders({type: ORDER_SELL, resourceType: SUBSCRIPTION_TOKEN})
        for (let x of tokens) {
            if (x.price <= Game.market.credits) {
                Game.market.deal(x.id, 1)
            }
        }
    }

    if ((Game.time - 20) % 10 === 0 && Memory.giveRoom) {
        Game.tools.give(Memory.giveRoom, 'energy', 2000000)
    }

    for (let rooms in Game.rooms) {
        let room = Game.rooms[rooms]
        if (room.controller && room.controller.my) {
            try {
                tower.work(room)
            } catch (e) {
                console.log(rooms + 'tower error ' + e)
            }

        }

    }

    Game.config.dofuncQueue()

    require('roomvisual').statistics()

}
module.exports.handlemission = function (roomNamein) {
    timer()
    for (let roomName in Memory.rooms) {
        if (roomNamein && roomName !== roomNamein) {
            continue
        }
        let room = Game.rooms[roomName]
        if (!room) continue
        mission_generator(room)
    }
    timer('mission_generator use=')
}
module.exports.handlespawn = function (roomName) {
    const room = Game.rooms[roomName]
    mission_spawner(room)
}
module.exports.spawncache = function () {
    mission_detector()
    for (let roomName in MissionCache) {
        console.log(roomName + JSON.stringify(MissionCache[roomName]))
    }
}